server/backup_runner.go

```diff
@@ -51,14 +66,85 @@ func (s *BackupRunner) Run(ctx context.Context, wg *sync.WaitGroup) {
 					}
 				}()
 				s.startAutoBackups(ctx, runningTasks, &mu)
+				s.downloadBinlogFiles(ctx)
 			}()
 		case <-ctx.Done(): // if cancel() execute
+			s.downloadBinlogTasks.wg.Wait()
+			return
+		}
+	}
+}
+
+func (s *BackupRunner) downloadBinlogFiles(ctx context.Context) {
+	normal := api.Normal
+	instanceFind := &api.InstanceFind{RowStatus: &normal}
+	instanceList, err := s.server.store.FindInstance(ctx, instanceFind)
+	if err != nil {
+		log.Error("Failed to retrieve instance list", zap.Error(err))
+		return
+	}
+
+	for _, instance := range instanceList {
+		if instance.Engine == db.MySQL && instance.RowStatus == api.Normal {
+			log.Debug("Checking if any databases of instance enables backup plan policy", zap.String("instance", instance.Name))
+			enabled, err := s.server.store.SomeDatabaseEnablesBackupPolicyInInstance(ctx, instance.ID)
+			if err != nil {
+				log.Error("Failed to get backup plan policy for databases of instance", zap.Error(err))
+				continue
+			}
+			if !enabled {
+				log.Debug("No database in instance enables backup plan policy, skip instance", zap.String("instance", instance.Name))
+				continue
+			}
+
+			s.downloadBinlogTasks.mu.Lock()
+			if _, ok := s.downloadBinlogTasks.running[instance.ID]; !ok {
+				s.downloadBinlogTasks.running[instance.ID] = true
+				go s.downloadBinlogFilesForInstance(ctx, instance, s.server.profile.DataDir, s.server.mysqlutil)
+				s.downloadBinlogTasks.wg.Add(1)
+			}
+			s.downloadBinlogTasks.mu.Unlock()
+		}
+	}
+}
+
+func (s *BackupRunner) downloadBinlogFilesForInstance(ctx context.Context, instance *api.Instance, dataDir string, mysqlutil mysqlutil.Instance) {
+	log.Debug("Downloading binlog files for MySQL instance", zap.String("instance", instance.Name))
+	defer func() {
+		s.downloadBinlogTasks.mu.Lock()
+		delete(s.downloadBinlogTasks.running, instance.ID)
+		s.downloadBinlogTasks.mu.Unlock()
+		s.downloadBinlogTasks.wg.Done()
+	}()
+	driver, err := getAdminDatabaseDriver(ctx, instance, "", "" /* pgInstanceDir */)
+	if err != nil {
+		if common.ErrorCode(err) == common.DbConnectionFailure {
+			log.Warn("Cannot connect to instance", zap.String("instance", instance.Name), zap.Error(err))
 			return
 		}
+		log.Error("Failed to get driver for MySQL instance when downloading binlog", zap.String("instance", instance.Name), zap.Error(err))
+		return
+	}
+	defer driver.Close(ctx)
+
+	binlogDir := getBinlogAbsDir(dataDir, instance.ID)
+	if err := createBinlogDir(dataDir, instance.ID); err != nil {
+		log.Error("Failed to create binlog directory", zap.Error(err))
+		return
+	}
+	mysqlDriver, ok := driver.(*mysql.Driver)
+	if !ok {
+		log.Error("Failed to cast driver to mysql.Driver", zap.String("instance", instance.Name))
+		return
+	}
+	mysqlDriver.SetUpForPITR(mysqlutil, binlogDir)
```

Comment:
nit: this is something we can optimize later.

Reply:
What do you mean by optimize?
It's MySQL specific, and we need to somehow set the fields before PITR.

Reply:
I mean to set up the binlogDir within the mysqlDriver without explicit set here. We can figure it out later, not this PR.

Reply:
SG.